### ISSUE TITLE
.NET Standard Support

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#8]
I don't think this is a blocking issue for NFig 3.0, but certainly a nice-to-have. We should look at supporting .NET Core 1.1 via the ".NET Standard 1.6".

I have no interest in supporting .NET Core 1.0 or `project.json` which is deprecated.

**Comment by bretcope** [comment]
Well, .NET is still a bit of a clusterfuck. It turns out what I was really waiting on was .NET Standard 2.0, and that is still tentatively set for Spring 2017.

In the meantime, the best option appears to be to support .NET Standard 1.3. According to [this table](https://blogs.msdn.microsoft.com/dotnet/2016/09/26/introducing-net-standard/#user-content-whats-new-in-net-standard-20), that will support both .NET Core 1.0 and .NET Framework 4.6. In my minimal testing, that appears to be true.

So I've started a [netstandard branch](https://github.com/NFig/NFig/tree/netstandard) which has the projects converted to the updated csproj style, and targeting .NET Standard 1.3. I was able to successfully create a nuget package from it, and install it in both .NET Core and .NET Framework console apps.

You'll likely need Visual Studio 2017 RC to open the solution now.

## AppVeyor

None of the automated build stuff on AppVeyor works yet because the way metadata works has completely changed. Most of the data that used to be in the nuspec and AssemblyInfo files now simply live in the csproj, which I'm overall a fan of. But it means the way I patch the version number at build has to change. Also, AppVeyor has to be configured to build using the 2017 tools, and to use msbuild to create the nuget package.

## Tests

There is still one problem: I haven't been able to get the tests to run yet. If I configure the Tests project to use .NET Standard, NUnit won't install (because it's not compatible yet). There is a preview Nuget package I can [get from here](https://www.myget.org/feed/nunit/package/nuget/NUnit) that installs, but I haven't been able to get the tests to actually run. ReSharper's test runner just ties, and I haven't had much luck getting the updated NUnit console runner to install. I could probably get it to work, but wow do I just hate dealing with these types of issues.

Another problem with .NET Standard is that there doesn't appear to be any encryption implementations which are compatible with it yet. So I can't test RSA or AES encrypted settings. I imagine this will change in the not too distant future, but it's annoying for now.

If I make the Tests project use .NET Framework, I end up with different problems. It doesn't accept the reference to a .NET Standard project, and gives me this error:

> Your project is not referencing the ".NETFramework,Version=v4.6" framework. Add a reference to ".NETFramework,Version=v4.6" in the "frameworks" section of your project.json, and then re-run NuGet restore.	

Of course, project.json isn't even a thing anymore, but I think that the error is trying to say is that the referenced project needs to target .NET Framework 4.6. It might work if we have an explicit .NET Framework build for NFig (separate csproj file), although that's something I was hoping to avoid.

I don't know what the best solution is at this point. @rossipedia if you have any ideas, I'd be happy to hear them.

**Comment by bretcope** [comment]
To follow up: it looks like the test frameworks are still trying to catch up with changes to .NET Standard. Both xUnit and NUnit are essentially broken for new development at the moment. Presumably that will get fixed in the next month or two now that VS2017 is officially out and the build tools are more stable. We probably still won't have NFig 3 released by then, but if this ever becomes the last blocker, I would say we could move on without it for now.

**Comment by bretcope** [comment]
An update to this. NUnit has an alpha release which now supports .NET Core and testing .NET Standard libraries: http://www.alteridem.net/2017/05/04/test-net-core-nunit-vs2017/

That was the major blocker for .NET Standard in my mind. We could also use xUnit, but I truly hate xUnit. Like, why can't they just call a test a "test"? Anyway, hopefully NUnit works and I won't have to rant about xUnit.

**Comment by bretcope** [comment]
The NFig project has been converted to .NET Standard 1.4. I think it could actually target as low as 1.3, but I haven't tried.

Additionally, I moved all of the build and versioning logic to [build.ps1](https://github.com/NFig/NFig/blob/master/build.ps1). This will make it easy to test locally, and even generate nuget packages locally for testing. Just run `.\build.ps1 -pack`. This greatly simplifies [appveyor.yml](https://github.com/NFig/NFig/blob/master/appveyor.yml).

The only thing I still haven't started working on is tests. As I said in the previous comment, NUnit _should_ work now with the current alpha. I just haven't explored it at all yet. We should be able to run the tests from Visual Studio and via the build script. Of course, I ripped out all of the old tests so the tests project is currently empty, but it would be easy to add in a few mock tests. @rossipedia do you have any interest in exploring getting NUnit hooked up?

**Comment by rossipedia** [comment]
Yeah, I can probably do that. NUnit 3 is fairly simple to hook into, although exactly _how_ we do that will depend on what scenarios we want to run the test (CI, GUI, Console, etc)

**Comment by rossipedia** [comment]
@bretcope does the test project need to be `netstandard1.4` as well? The latest NUnit looks like it needs at least `netstandard1.6`:

```
PM> Install-Package -ProjectName NFig.Tests -Id NUnit
...
Restoring packages for C:\Code\NFig\NFig.Tests\NFig.Tests.csproj...
Install-Package : Package NUnit 3.6.1 is not compatible with netstandard1.4 (.NETStandard,Version=v1.4). Package NUnit 3.6.1 supports:
  - monoandroid (MonoAndroid,Version=v0.0)
  - net20 (.NETFramework,Version=v2.0)
  - net35 (.NETFramework,Version=v3.5)
  - net40 (.NETFramework,Version=v4.0)
  - net45 (.NETFramework,Version=v4.5)
  - netstandard1.6 (.NETStandard,Version=v1.6)
  - portable-net45+win8+wp8+wpa81 (.NETPortable,Version=v0.0,Profile=Profile259)
  - xamarinios10 (Xamarin.iOS,Version=v1.0)
At line:1 char:1
```

**Comment by rossipedia** [comment]
Ugh... looks like it needs to be `netstandard*`:
```
1>------ Build started: Project: NFig, Configuration: Debug Any CPU ------
1>NFig -> C:\Code\NFig\NFig\bin\Debug\netstandard1.4\NFig.dll
2>------ Build started: Project: NFig.Tests, Configuration: Debug Any CPU ------
2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.Common.targets(73,5): error : 
    Project 'C:\Code\NFig\NFig\NFig.csproj' targets '.NETStandard,Version=v1.4'. 
    It cannot be referenced by a project that targets '.NETFramework,Version=v4.6'.
```

**Comment by bretcope** [comment]
I don't think it matters too much what the Tests project targets since it doesn't get distributed.

**Comment by rossipedia** [comment]
Agreed. I'm just trying to get _something_ working

**Comment by bretcope** [comment]
I added the first real tests and it blew up complaining that it couldn't find System.Reflection.TypeExtensions. [See build log](https://ci.appveyor.com/project/bretcope/nfig/build/141#L129). I got the same result locally.

It appears that when you reference a .NET Standard project from a .NET Framework project, it doesn't pull in the nuget dependencies of the .NET Standard library. The workaround is to manually install the nuget dependencies in the Tests project. It's a bit annoying that we have to maintain that, but I'm not sure if there's an alternative at this point.

If you happen to have any ideas @rossipedia, I'm open to them.

**Comment by bretcope** [comment]
Well, that workaround didn't actually work. It just failed on a different library. I think the issue is that, while all of the dependencies are installed, some of the dlls are for the wrong target, and minor incompatibilities are going to surface.

The easiest option, for the moment appears to be to add .NET Framework 4.6.1 as a target for NFig. It's probably not such a bad thing to have that target anyway. I should do a little bit of cleanup work on the references for that target since the Framework target won't need as many nuget dependencies.

Right now, only .NET Framework is really getting tested. Longer term, I think we want to be able to run tests on both Framework and Core. If that's a pain to implement with NUnit right now, we could wait until the tooling catches up.

**Comment by bretcope** [comment]
Oh, another option I considered was to pack NFig as a post-build step and have it install into NFig.Tests. That would probably be the most realistic way to test it, but also a bit clunky. Our testing strategy will probably continue to evolve as we go.
